src/Data/TypeRep/Map.hs

```diff
@@ -0,0 +1,40 @@
+{-# LANGUAGE ExplicitForAll      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Data.TypeRep.Map
+       ( TypeRepMap (..)
+       , empty
+       , insert
+       , lookup
+       ) where
+
+import Prelude hiding (lookup)
+
+import Control.Monad ((<=<))
+import Data.Dynamic (Dynamic, fromDynamic, toDyn)
+import Data.Proxy (Proxy (..))
+import Data.Typeable (TypeRep, Typeable, typeOf, typeRep)
+
+import qualified Data.Map.Lazy as LMap
+
+-- | Map-like data structure with types served as the keys.
+newtype TypeRepMap = TypeRepMap
+    { unMap :: LMap.Map TypeRep Dynamic
+    } deriving (Show)
+
+-- | Empty structure.
+empty :: TypeRepMap
+empty = TypeRepMap mempty
+
+-- | Inserts the value with its type as a key.
+insert :: Typeable a => a -> TypeRepMap -> TypeRepMap
+insert val = TypeRepMap . LMap.insert (typeOf val) (toDyn val) . unMap
+
+-- | Looks up the value at the type.
+-- >>> let x = lookup $ insert (11 :: Int) empty
+-- >>> x :: Maybe Int
+-- Just 11
+-- >>> x :: Maybe ()
+-- Nothing
+lookup :: forall a . Typeable a => TypeRepMap -> Maybe a
+lookup = fromDynamic <=< LMap.lookup (typeRep (Proxy :: Proxy a)) . unMap
```

Comment:
When `fromDynamic` returns `Nothing` here, this means a bug in our implementation, but with `<=<` this can go unnoticed to the user (he just receives `Nothing`). Instead we should throw in this case.

(But don't bother making this change, because this code will soon change in a way that `Dynamic` is not used, I'm just pointing this out)